File: leetcode-04/solutions/python3/0133.py
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return None
        if node in self.dict:
            return self.dict[node]

        self.dict[node] = Node(node.val, [])
        for neighbor in node.neighbors:
            self.dict[node].neighbors.append(self.cloneGraph(neighbor))

        return self.dict[node]

    dict = {}

File: leetcode-04/solutions/cpp/0133.cpp
class Solution {
 public:
  Node* cloneGraph(Node* node) {
    if (!node) return NULL;
    if (map.count(node)) return map[node];

    map[node] = new Node(node->val, {});
    for (Node* neighbor : node->neighbors)
      map[node]->neighbors.push_back(cloneGraph(neighbor));

    return map[node];
  }

 private:
  unordered_map<Node*, Node*> map;
};